File: analysis/src/extract_gt_methods.cpp
#include "json_loader.h"
#include <set>
#include <sstream>

static constexpr int kBaseAddr{0x400000};

int main(int argc, char *argv[]) {
  if (argc != 2) {
    std::cerr << "usage: ./extract_gt_methods <path/to/json>" << std::endl;
    return EXIT_FAILURE;
  }

  boost::json::object structures =
      JsonLoader::LoadData(argv[1]).as_object()["structures"].as_object();

    // Put in set to sort
    std::set<int> method_addrs;
  for (auto &cls : structures) {
    for (auto &method : cls.value().as_object()["methods"].as_object()) {
        std::string ea_str = method.value().as_object()["ea"].as_string().c_str();
        std::stringstream ss;
        ss << std::hex << ea_str;
        int out;
        ss >> out;
        out -= kBaseAddr;
        method_addrs.insert(out);
    }
  }

    // print sorted set 
  for (int ii : method_addrs) {
    std::cout << ii << std::endl;
  }

  return 0;
}

File: analysis/extract_gt_methods.py
import json5
import sys
import os
import pathlib

fpath = pathlib.Path(__file__).parent.absolute()

sys.path.append(os.path.join(fpath, '..'))

from parseconfig import config

kBaseAddr = 0x400000

def main():
    json_file = config['gtResultsJson']

    structures = json5.load(open(json_file, 'r'))['structures']

    method_addrs = set()

    for cls in structures.values():
        for method in cls['methods'].values():
            ea = method['ea']
            method_addrs.add(int(ea, 16) - kBaseAddr)

    with open(config['gtMethodsPath'], 'w') as f:
        for method in method_addrs:
            f.write(str(method) + '\n')

if __name__ == '__main__':
    main()
